### PR TITLE
Cucumber Expressions for Python

### LINKED ISSUE #27: Python implementation of Cucumber Expressions
The [Behave](https://github.com/behave/behave) team (led by @jenisys) would like to add support for Cucumber Expressions in Behave.

The first step would be to implement Cucumber Expressions in Python.
There are four existing implementations in this repository (Go, Ruby, TypeScript and Java). The Python implementation should follow the structure of the existing implementations as closely as possible. This will make maintenance easier in the future.

Every implementation uses a shared set of [test data](https://github.com/cucumber/cucumber-expressions/tree/main/testdata), which are YAML files with input and expected output for different parts of the parser. 

The parser consists of multiple components, and it may be easier to implement the library in the following order:

* `CucumberExpressionTokenizer`
* `CucumberExpressionParser`
* `TreeRegexp`
* `CucumberExpression`
* `CucumberExpressionGenerator`

I would also recommend using TDD to guide the implementation. That is - start by writing the test for a component (reading in thest data from the relevant YAML file). Make the tests pass one by one, by making incremental improvements to the code.

If contributors are working in a team, it may be easier to collaborate if all contributors are committing to the same feature branch in the same repo (as opposed to using individual forks). Please add comment here if you want commit access so you can work on a branch in this repo.

### PR DESCRIPTION
**Pull request by jsa34**
# Description

Cucumber expressions for Python

# Motivation & context

Fixes #27 

## Type of change

- New feature (non-breaking change which adds new behaviour)

# Checklist:

- [X] I have read the [**CONTRIBUTING**](../CONTRIBUTING.md) document.
- [X] My change needed additional tests
  - [X] I have added tests to cover my changes.
- [X] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

### CODE REVIEW

**Comment by jsa34** [comment]
I may have missed some bits, but wanted to get the first "draft" out for feedback.

I can't work out the only test that fails: test_prefers_expression_with_longest_non_empty_match

**Comment by jenisys** [comment]
I just look over your pull-request (not the contents, just the filename).

You violated the PEP-8 naming conventions for Python by using CamelCase instead of snake_case for Python packages/modules.

* python/Cucumber/CucumberExpressions/ => SHOULD BE: python/cucumber_expressions/

  REASONS:
  * Overly and unnecessarily complex
  * Naming style: You have a "tautology" (you name "cucumber" twice during imports; "Doppelt gemoppelt)
  * UNVERIFIED: Your "Cucumber" is probably not a Python namespace module. Therefore, it will block others from using the "Cucumber" prefix/package-name.

SEE ALSO:

* [PEP-0008: Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/)

OTHERWISE:

* Unit tests within the package: I would normally move them out of the package

  REASONS:
  * The installed package is smaller (less "FAT").
  * The installed tests within the installed package are nearly never used.

**Comment by jsa34** [comment]
Great feedback, @jenisys - thanks!

I am not an expert in Python so thought I would pop what I have done so far rather than getting too far down the line and not realising my mistakes.

I will have a look at what you have suggested - having never released any artifacts for python, all that area and nomenclature are new territory.

Many thanks again!

**Comment by jenisys** [comment]
No problem, learning by doing.
Keep up your good work 😀

**Comment by blaisep** [comment]
Outstanding effort @jsa34 , thanks so much for contributing. I rounded up some pythonistas to help with this effort just before the holidays and I think everyone was too busy to take action.
@mattwynne suggested that we try to make the python port adhere to the "pythonic" conventions and it's wonderful that @jenisys provides guidance.

I would be very happy to work with you if you are interested in any kind of collaboration.

> Great feedback, @jenisys - thanks!
> 
> I am not an expert in Python so thought I would pop what I have done so far rather than getting too far down the line and not realising my mistakes.
> 
> I will have a look at what you have suggested - having never released any artifacts for python, all that area and nomenclature are new territory.
> 
> Many thanks again!

**Comment by jsa34** [comment]
> Outstanding effort @jsa34 , thanks so much for contributing. I rounded up some pythonistas to help with this effort just before the holidays and I think everyone was too busy to take action. @mattwynne suggested that we try to make the python port adhere to the "pythonic" conventions and it's wonderful that @jenisys provides guidance.
> 
> I would be very happy to work with you if you are interested in any kind of collaboration.
> 
> > Great feedback, @jenisys - thanks!
> > I am not an expert in Python so thought I would pop what I have done so far rather than getting too far down the line and not realising my mistakes.
> > I will have a look at what you have suggested - having never released any artifacts for python, all that area and nomenclature are new territory.
> > Many thanks again!

Hello!

That's great and would be very much appreciated.

I tried to keep as close to the other lang implementations as possible with a sprinkling of pythonisms where possible, but where the line is drawn I'm not sure as the closer it is to Ruby/JS, etc I guess maintainability for cross language feature is better, but again I deliberately thought I would share at this point so I can start discussions and gather feedback. 

I would very much appreciate any support/collaboration as it will most definitely help me learn more about Python and hopefully help me become more "pythonic"!

Let me know what you suggest :)

**Comment by blaisep** [comment]
@jsa34 , I would say let's go ahead and `merge main` into your PR so that you can resolve the merge conflicts (a courtesy to @jenisys ).

I think that @mattwynne is correct is wanting to make the language-specific implementations representative of the native customs.

**Comment by jsa34** [comment]
Resolved merge conflicts.

Happy to make this reflect more the language native customs, but if you have any able to provide any guidance/examples of this that would be gratefully appreciated!

Happy to chat about it or to go through code comments

File: CHANGELOG.md
Comment (aslakhellesoy) [inline CHANGELOG.md:19]: Please incluse a Markdown link to this PR

File: README.md
Comment (aslakhellesoy) [inline README.md:92]: Please remove `{{% stepdef-body %}}` - that's a leftover from an abaondoned documentation tool.

Comment (aslakhellesoy) [reply README.md:92]: Why did you choose to name it `param_type`? I don't think `type` is a reseved keyword in Python.

File: python/.gitignore
Comment (aslakhellesoy) [inline python/.gitignore:5]: Please remove irrelevant entries from this file

File: python/.pre-commit-config.yaml
Comment (aslakhellesoy) [inline python/.pre-commit-config.yaml:1]: What is this file for? Can we remove it?

**Review by aslakhellesoy — CHANGES_REQUESTED** [review]
Excellent work @jsa34 - thank you!

Could you please add build scripts and/or instructions so other contributors can run all tests with a single command after cloning the repo?

- Maybe add a `requirements.txt` file that lists dependencies
- How to run all the tests

Could you also please add a GitHub Action that does all of the above? You can look at this one for inspiration: https://github.com/cucumber/tag-expressions/blob/main/.github/workflows/test-python.yml

@jenisys I'm not sure it's a good idea to rely on make for this (Windows users don't always have it). What's the most idiomatic and cross platform tool in Python to write build scripts?

File: README.md
Comment (jsa34) [reply README.md:92]: Will address point 1.

RE point 2, I don't think it's "wrong" per se, but I thought it was bad practice to use type as a func param.  Happy to stand corrected if any pythonistas correct me, though!

See: https://stackoverflow.com/questions/10568087/is-it-safe-to-use-the-python-word-type-in-my-code  

Happy to review this accordingly

File: python/.gitignore
Comment (jsa34) [reply python/.gitignore:5]: Apologies - I popped in my local gitignore that I use as a standard template.  Will tidy up

File: CHANGELOG.md
Comment (jsa34) [reply CHANGELOG.md:19]: > 

1. No dependencies are needed, apart from testing (so I put the requirements in the test dir.  Happy to change this - only 2 dependencies needed for running tests: pytest and pyyaml (to read test yaml files).  Actual modules has 0 dependencies.
2. I will write a doc for running tests - good point!
3. I have never used Github actions, but will give it a go!

**Comment by jsa34** [comment]
Successfully (well, functionally) implemented the github actions.

It highlighted there were some compat issues with previous versions of python.  I have used python 3.7 as the baseline version as this is the latest version currently supported by Python.

I added the extra features that were committed for other langs for extra data types so that it is compatible with the other languages still.

So, just need to know if usage of "type" as a param (which I have avoided)

File: python/.pre-commit-config.yaml
Comment (aslakhellesoy) [reply python/.pre-commit-config.yaml]: Actually, I think enforcing coding standards is a good thing! If we keep the file, can we have CI run the check using this tool?

File: CHANGELOG.md
Comment (aslakhellesoy) [reply CHANGELOG.md:19]: Thanks - nice to see all the tests passing in CI!

File: python/cucumber_expressions/parameter_type.py
Comment (aurelien-reeves) [inline python/cucumber_expressions/parameter_type.py:45]:
```suggestion
    # @param param_type the return type of the transformed
```

**Review by aurelien-reeves — COMMENTED** [review]
Note: I am not a python developer: I did not review the Python code

It looks pretty good. Good test coverage, good documentation, acceptance tests seems to be passing too, and the CI is testing all the things 😁

Apart from the question regarding the name of the `type` ParameterType parameter, it looks good to me.

For that `type` thing, here's my humble opinion: `type` is actually part of the DSL of cucumber expressions. And its usage is pretty moderated in the code. I would keep it `type`.

What about keeping `type` for the public API, but using `param_type` in the private code?

But if it is strongly a bad thing in python, I am good with `param_type`.

File: python/cucumber_expressions/parameter_type.py
Comment (jsa34) [reply python/cucumber_expressions/parameter_type.py]: I have changed param_type -> type, with the caveat that I'm not 100% sure of the consequences, but the tests were fine...

File: python/.pre-commit-config.yaml
Comment (jsa34) [reply python/.pre-commit-config.yaml:1]: I've reinstated the file.

I'll take a look, but I know this is aimed at running checks before a git commit.

I have a feeling I saw something about a pre-commit-ci that I think works with github actions, which would be great.

Comment (jsa34) [reply python/.pre-commit-config.yaml:1]: Looks like the best approach is using https://pre-commit.ci/ which is available on the github marketplace, for what I can see, and is free for open source projects... A nice to have, perhaps?

Comment (aurelien-reeves) [reply python/.pre-commit-config.yaml:1]: I guess @aslakhellesoy suggested to check styling/standards as part of the CI, but not using a pre-commit hook

Like it is done for example in javascript with eslint, or ruby with rubocop: those are part of the CI beside the tests

**Review by aurelien-reeves — COMMENTED** [review]
It looks really good to me.
This is an amazing work, properly tested and documented. I give it a go.

There's just that pre-commit hook thing (see below)

**Comment by aurelien-reeves** [comment]
> > It looks really good to me. This is an amazing work, properly tested and documented. I give it a go.
> > There's just that pre-commit hook thing (see below)
> 
> Ah, apologies, I misunderstood - I thought from @aslakhellesoy's comments he talked about keeping the file and running in the CI.
> 
> I find pre-commit great because it gatekeeps before reaching a commit, but happy to pop in lint checks explicitly. There does appear to be an ecosystem for using pre-commit files as above, though, which I thought would be the best approach - as I say, happy either way!

I have no opinion on pre-commit hooks.
it is just that at the moment, at Cucumber, there are no use of those.

That may be interesting, later, why not. But that could be part of a brand new PR which would bring those at the root of the repo, or for all languages at once :)

**Comment by jsa34** [comment]
> > > It looks really good to me. This is an amazing work, properly tested and documented. I give it a go.
> > > There's just that pre-commit hook thing (see below)
> > 
> > 
> > Ah, apologies, I misunderstood - I thought from @aslakhellesoy's comments he talked about keeping the file and running in the CI.
> > I find pre-commit great because it gatekeeps before reaching a commit, but happy to pop in lint checks explicitly. There does appear to be an ecosystem for using pre-commit files as above, though, which I thought would be the best approach - as I say, happy either way!
> 
> I have no opinion on pre-commit hooks. it is just that at the moment, at Cucumber, there are no use of those.
> 
> That may be interesting, later, why not. But that could be part of a brand new PR which would bring those at the root of the repo, or for all languages at once :)

👍 
For the moment, I have just enabled the precommit as a linting stage as you described for ruby and javascript, but can be used locally as a git hook, too - best of both worlds!

**Comment by aurelien-reeves** [comment]
I won't merge too early.
Not being a python dev, I want to let some time to other reviewers to make sure I did not miss something

**Comment by jenisys** [comment]
@aslakhellesoy (question from review above):

> @jenisys I'm not sure it's a good idea to rely on make for this (Windows users don't always have it). What's the most 
> idiomatic and cross platform tool in Python to write build scripts?

Correct, using "make" normally does not work on Windows for many users.
I normally used [invoke](https://www.pyinvoke.org) (https://github.com/pyinvoke/invoke), as a small build tool.

Otherwise, if a "pyproject.toml" is provide (newer mechanism instead of "setup.py") to be able to install the package, 
you can use as build script for developer tasks (with: `poetry` or ...; I haven't used it much yet).

**BUT:** Just for testing it, a simpler solution would be a short description in the README.

```bash
$ cd cucumber-expressions/python/

# -- BETTER: Use "requirements.txt" (or: "py.requirements.txt")
$ pip install pytest coverage PyYAML

# -- Run the tests:
$ pytest

# -- Run test with code coverage:
$ coverage run -m pytest

# -- AFTERWARDS:
$ coverage report
...
```

REVIEW NOTE @jsa34:
* I am currently missing a mechanism to install the package, like: "setup.py" file or "pyproject.toml" file !?!
* I will look into the other question above (if the mechanism is pythonic).

File: python/cucumber_expressions/parameter_type.py
Comment (jenisys) [reply python/cucumber_expressions/parameter_type.py]: @jsa34 @aurelien-reeves 
Using `type` as parameter name normally causes problems/warnings with static code analysis tools, like `pylint`.
REASON: `type()` is a builtin function in Python to determine the type of an object.

Therefore, using the `param_type` as parameter name is probably the better choice.
NOTE: I still need to look over the complete implementation as whole.

**Comment by jenisys** [comment]
NOTE: The `cucumber-expressions/python/test/README.md` contains a description how to run the tests (like my short description above). Maybe this description should be moved up one level into the `cucumber-expressions/python/README.md`.

**Comment by aslakhellesoy** [comment]
I think this is good to go now. Any objections if we merge it?

We still need a GA workflow for releasing the Python code, but I think that should be done separately from this PR.

**Comment by jenisys** [comment]
@aslakhellesoy @jsa34
There are two options:

* Merge it now and continue or
* Provide some fixes and merge then

NOTES:
* I ran "pylint" over the code yesterday and it stumbled over a circular dependency problems and some other issues (besides the noise that docsting/documentation is missing)
* Currently, a method to install the package in Python is missing: No "setup.py" or "pyproject.toml"
* I try to look over the remaining stuff tonight

**Review by aslakhellesoy — APPROVED** [review]
Great work!! Just waiting for @jenisys' final changes/feedback now.

**Comment by blaisep** [comment]
@jsa34 , Sorry I have been AFK

> Let me know what you suggest :)

This is really very nice work!!

**Comment by jsa34** [comment]
I have responded to @blaisep 's feedback.  I already had updated the README.md entry (I hope!) - thanks, @blaisep !

I hadn't considered the build part in this scope, and have never had to release anything, but just learned about Poetry and seemed like a great idea to use (learning so much doing this!)  I have updated this to use poetry and so if this accepted I *believe* adding the deployment script should be trivial, but never released anything on Pypi so thought I would hold fire going that far :)

**Comment by jenisys** [comment]
@aslakhellesoy @jsa34 
Ok, I had a short look over the current state (not a complete review).
Most of the problems/smells from the "pylint" run seem to by solved by the latest changes from @jsa34 .

NECESSARY BEFORE THE MERGE (in my opinion):

* [x] "pyproject.toml" exists now
* [ ] P1
* [ ] P2
* [ ] OPTIONAL: P3 should be discussed or fixed

DETAILS:

- P1. Running pytest with Python 3.9 on my Mac, I get a DeprecationWarning on:
  `cucumber_expressions/cucumber_expression.py:19`
  `ESCAPE_PATTERN = b"([\\^\[({$.|?*+})\]])"  # noqa W605`
  SHOULD BE: 
  `ESCAPE_PATTERN = rb"([\\^\[({$.|?*+})\]])"  # noqa W605`

  HINT:
  - That fixes also the "pylint" warning(s)
  - Regular expression checks (what is complained about and what not) has changed in the last few Python versions.
    I think it started with Python 3.9, and continued with Python 3.10.

- P2. BAD_DOCSTRING: docstring is in the wrong position:
  Class docstring position is used for the description of the Class ctor (`__init__()`)
  EXAMPLES:
    - `cucumber_expressions.parameter_type:ParameterType`
    - `cucumber_expressions.regular_expression:RegularExpression`
    - ...

- P3. NAMING: TAUTOLOGY (naming is unnecessarily complex; same name is repeated on different levels):

  * RENAME: "cucumber_expressions/cucumber_expression.py" to "cucumber_expressions/expression.py"
  * RENAME: "cucumber_expressions/cucumber_expression_generator.py" to "cucumber_expressions/expression_generator.py"
  * RENAME: "cucumber_expressions/cucumber_expression_parser.py" to "cucumber_expressions/expression_parser.py"
  * RENAME: "cucumber_expressions/cucumber_expression_tokenizer.py" to "cucumber_expressions/expression_tokenizer.py"

- P4. BAD-NAME: Argument name "at" -- BAD NAME (in my opinion), BETTER: pos (or: position)
  -- Meant is an index/position within a list.

- P5. MINOR: Module structure is not really Pythonic.
  SMELL: Each module contains just one class with the same name as module, just in CamelCase.
  EXPLANATION: In Python, the module structure is normally not so fine grained.
  In addition, you have multiple modules that seem to form a group (have same prefix).
  Therefore, these seem to belong together (on first sight).

- P6. MINOR: Constant regexp patterns are not compiled
  HINT: Constant regular expression patterns are faster to use if you compile them.
  AFFECTED: cucumber_expressions.parameter_type:ParameterType, ...
  NAMING: Same "regexp" is used for different things:
  The regexp-pattern (as string) and the regexp-object (as fast matcher)

- P7. COULD BE IMPROVED: Documentation of source code (docstring): Nearly non-existing.
  => Should be improved in the future, especially for module- and class-descriptions.

**Comment by blaisep** [comment]
@jsa34 ,
>  but never released anything on Pypi so thought I would hold fire going that far :)
Many pythonistas [use `flit` to package modules for pypi](https://flit.readthedocs.io/en/latest/rationale.html) because it's almost plug n play ... so you might want to consider that to get started.

**Comment by jsa34** [comment]
Hello!

Thank you once again for great feedback.

I have address what I can, but I agree the documentation needs to be looked properly, which would be good to come back to with a follow up MR as I don't have capacity at the moment.

I have added a release yml (provisionally), but it obviously needs secrets set up.

Any other feedback, as always, much appreciated!

**Comment by blaisep** [comment]
Hi @jsa34 , my schedule is more open this week, and I would be very happy to help with the documentation as well as to borrow some of [the cleverness @jenisys does with `invoke`](https://github.com/behave/behave/tree/master/tasks)

It may be easier to do this realtime. You can reach me at home: blaise at gmail dot com
or drop time on my calendar.
I do a lot of pair and mob programming, [for example](https://www.twitch.tv/videos/1278857586)


> I have address what I can, but I agree the documentation needs to be looked properly, which would be good to come back to with a follow up MR as I don't have capacity at the moment.

**Comment by aslakhellesoy** [comment]
Is anything holding this PR up? If there are remaining issues to be fixed, can we do that in a new PR?

**Comment by jenisys** [comment]
AFAIK: no.

Any remaining things can be done afterwards (if necessary).

**Pull request merged**
